
<!DOCTYPE HTML>

<html lang="en" class="_theme-default">

<head>
  <meta charset="utf-8">
  <title>Catch - PHP - W3cubDocs</title>
  
  <meta name="description" content="PHP 5 has an exception model similar to that of other programming languages. An exception can be thrown, and caught (&#34;catched&#34;) within PHP. &hellip;">
  <meta name="keywords" content="exceptions, catch, php">
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="https://docs.w3cub.com/php/language.exceptions.html">
  <link href="/favicon.png" rel="icon">
  <link rel="stylesheet" type="text/css" href="/assets/application-01fda2ddb8339756caccf7add5ad4cf849ab52d069bd799015c7f04f93164f64753bff0d15a49d8060b1e66e41002bb301ccadc2350937df079cea3cd52d3cca.css">
  <script src="/assets/application-d9be6f56a823612443fc15b2e027a630e02c4ad2685bb750d13fa4fae28d46c3e7f7ebb69bd4bafddf116f218f9372e9be44021d4247dc20424e2fd1ff8cef81.js" type="text/javascript"></script>
  <script src="/json/php.js"></script>
  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-71174418-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>
  <script data-ad-client="ca-pub-2572770204602497" async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
  <script async custom-element="amp-auto-ads"
  src="https://cdn.ampproject.org/v0/amp-auto-ads-0.1.js">
</script>


</head>

<body class="docs">
	<amp-auto-ads type="adsense"
              data-ad-client="ca-pub-2572770204602497">
	</amp-auto-ads>
	<div class="_app">
	<header class="_header">

  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/php/" class="_nav-link" title="" style="margin-left:0;">PHP</a></span>
  
  <nav class="_nav">
    <a href="https://tools.w3cub.com/?_sp=docs" target="_blank" class="_nav-link ">W3cubTools</a>
    <a href="/cheatsheets/" class="_nav-link ">Cheatsheets</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		
		<form class="_search">
		  <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
		  <a class="_search-clear"></a>
		  <div class="_search-tag"></div>
		</form>
		
		<div class="_list-wrap">
			<div class="_list">
			
			</div>
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="6861657091"
     data-ad-format="link"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
			<div class="_page _php">
				
				
<h1>Exceptions</h1> <h2>Table of Contents</h2>
<ul class="chunklist chunklist_chapter"><li><a href="language.exceptions.extending">Extending Exceptions</a></li></ul> <div class="simplesect"> <p class="para"> PHP 5 has an exception model similar to that of other programming languages. An exception can be <a href="language.exceptions" class="link"><em>throw</em></a>n, and caught ("<a href="language.exceptions#language.exceptions.catch" class="link"><em>catch</em></a>ed") within PHP. Code may be surrounded in a <a href="language.exceptions" class="link"><em>try</em></a> block, to facilitate the catching of potential exceptions. Each <a href="language.exceptions" class="link"><em>try</em></a> must have at least one corresponding <a href="language.exceptions#language.exceptions.catch" class="link"><em>catch</em></a> or <a href="language.exceptions#language.exceptions.finally" class="link"><em>finally</em></a> block. </p> <p class="para"> The thrown object must be an instance of the <a href="class.exception" class="classname">Exception</a> class or a subclass of <a href="class.exception" class="classname">Exception</a>. Trying to throw an object that is not will result in a PHP Fatal Error. </p> </div> <div class="simplesect"> <h3 class="title"><em>catch</em></h3> <p class="para"> Multiple <a href="language.exceptions#language.exceptions.catch" class="link"><em>catch</em></a> blocks can be used to catch different classes of exceptions. Normal execution (when no exception is thrown within the <a href="language.exceptions" class="link"><em>try</em></a> block) will continue after that last <a href="language.exceptions#language.exceptions.catch" class="link"><em>catch</em></a> block defined in sequence. Exceptions can be <a href="language.exceptions" class="link"><em>throw</em></a>n (or re-thrown) within a <a href="language.exceptions#language.exceptions.catch" class="link"><em>catch</em></a> block. </p> <p class="para"> When an exception is thrown, code following the statement will not be executed, and PHP will attempt to find the first matching <a href="language.exceptions#language.exceptions.catch" class="link"><em>catch</em></a> block. If an exception is not caught, a PHP Fatal Error will be issued with an "<em>Uncaught Exception ...</em>" message, unless a handler has been defined with <span class="function"><a href="function.set-exception-handler" class="function">set_exception_handler()</a></span>. </p> <p class="para"> In PHP 7.1 and later, a <a href="language.exceptions#language.exceptions.catch" class="link"><em>catch</em></a> block may specify multiple exceptions using the pipe (<em>|</em>) character. This is useful for when different exceptions from different class hierarchies are handled the same. </p> </div> <div class="simplesect"> <h3 class="title"><em>finally</em></h3> <p class="para"> In PHP 5.5 and later, a <a href="language.exceptions#language.exceptions.finally" class="link"><em>finally</em></a> block may also be specified after or instead of <a href="language.exceptions#language.exceptions.catch" class="link"><em>catch</em></a> blocks. Code within the <a href="language.exceptions#language.exceptions.finally" class="link"><em>finally</em></a> block will always be executed after the <a href="language.exceptions" class="link"><em>try</em></a> and <a href="language.exceptions#language.exceptions.catch" class="link"><em>catch</em></a> blocks, regardless of whether an exception has been thrown, and before normal execution resumes. </p> </div> <div class="simplesect"> <h3 class="title">Notes</h3> <blockquote class="note">
<p><strong class="note">Note</strong>: </p>
<p class="para"> Internal PHP functions mainly use <a href="https://secure.php.net/manual/en/errorfunc.configuration.php#ini.error-reporting" class="link">Error reporting</a>, only modern <a href="https://secure.php.net/manual/en/language.oop5.php" class="link">Object oriented</a> extensions use exceptions. However, errors can be simply translated to exceptions with <a href="class.errorexception" class="link">ErrorException</a>. </p> </blockquote> <div class="tip">
<strong class="tip">Tip</strong> <p class="para"> The <a href="https://secure.php.net/manual/en/intro.spl.php" class="link">Standard PHP Library (SPL)</a> provides a good number of <a href="https://secure.php.net/manual/en/spl.exceptions.php" class="link">built-in exceptions</a>. </p> </div> </div> <div class="simplesect"> <h3 class="title">Examples</h3> <div class="example" id="example-280"> <p><strong>Example #3 Throwing an Exception</strong></p> <div class="example-contents"> <pre class="phpcode" data-language="php">&lt;?php
function inverse($x) {
    if (!$x) {
        throw new Exception('Division by zero.');
    }
    return 1/$x;
}

try {
    echo inverse(5) . "\n";
    echo inverse(0) . "\n";
} catch (Exception $e) {
    echo 'Caught exception: ',  $e-&gt;getMessage(), "\n";
}

// Continue execution
echo "Hello World\n";
?&gt;</pre> </div> <div class="example-contents"><p>The above example will output:</p></div> <div class="example-contents screen"> <div class="cdata"><pre>
0.2
Caught exception: Division by zero.
Hello World
</pre></div> </div> </div> <div class="example" id="example-281"> <p><strong>Example #4 Exception handling with a <em>finally</em> block</strong></p> <div class="example-contents"> <pre class="phpcode" data-language="php">&lt;?php
function inverse($x) {
    if (!$x) {
        throw new Exception('Division by zero.');
    }
    return 1/$x;
}

try {
    echo inverse(5) . "\n";
} catch (Exception $e) {
    echo 'Caught exception: ',  $e-&gt;getMessage(), "\n";
} finally {
    echo "First finally.\n";
}

try {
    echo inverse(0) . "\n";
} catch (Exception $e) {
    echo 'Caught exception: ',  $e-&gt;getMessage(), "\n";
} finally {
    echo "Second finally.\n";
}

// Continue execution
echo "Hello World\n";
?&gt;</pre> </div> <div class="example-contents"><p>The above example will output:</p></div> <div class="example-contents screen"> <div class="cdata"><pre>
0.2
First finally.
Caught exception: Division by zero.
Second finally.
Hello World
</pre></div> </div> </div> <div class="example" id="example-282"> <p><strong>Example #5 Nested Exception</strong></p> <div class="example-contents"> <pre class="phpcode" data-language="php">&lt;?php

class MyException extends Exception { }

class Test {
    public function testing() {
        try {
            try {
                throw new MyException('foo!');
            } catch (MyException $e) {
                // rethrow it
                throw $e;
            }
        } catch (Exception $e) {
            var_dump($e-&gt;getMessage());
        }
    }
}

$foo = new Test;
$foo-&gt;testing();

?&gt;</pre> </div> <div class="example-contents"><p>The above example will output:</p></div> <div class="example-contents screen"> <div class="cdata"><pre>
string(4) "foo!"
</pre></div> </div> </div> <div class="example" id="example-283"> <p><strong>Example #6 Multi catch exception handling</strong></p> <div class="example-contents"> <pre class="phpcode" data-language="php">&lt;?php

class MyException extends Exception { }

class MyOtherException extends Exception { }

class Test {
    public function testing() {
        try {
            throw new MyException();
        } catch (MyException | MyOtherException $e) {
            var_dump(get_class($e));
        }
    }
}

$foo = new Test;
$foo-&gt;testing();

?&gt;</pre> </div> <div class="example-contents"><p>The above example will output:</p></div> <div class="example-contents screen"> <div class="cdata"><pre>
string(11) "MyException"
</pre></div> </div> </div> </div> <div class="manualnavbar" style="text-align: center;"> <div class="prev" style="text-align: left; float: left;"><a href="language.errors.php7">← Errors in PHP 7</a></div> <div class="next" style="text-align: right; float: right;"><a href="language.exceptions.extending">Extending Exceptions →</a></div>   </div>
<div class="_attribution">
  <p class="_attribution-p">
    © 1997–2018 The PHP Documentation Group<br>Licensed under the Creative Commons Attribution License v3.0 or later.<br>
    <a href="https://secure.php.net/manual/en/language.exceptions.php" class="_attribution-link">https://secure.php.net/manual/en/language.exceptions.php</a>
  </p>
</div>

				
			</div>
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="1992473792"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
		</div>
	</section>

	</div>
	<svg style="display:none">
		<symbol id="icon-dir"><svg viewBox="0 0 20 20"><path d="M15 10c0 .3-.305.515-.305.515l-8.56 5.303c-.625.41-1.135.106-1.135-.67V4.853c0-.777.51-1.078 1.135-.67l8.56 5.305S15 9.702 15 10z"/></svg></symbol>
	  </svg>
</body>
</html>
